cpp/src/arrow/compute/exec/exec_plan.cc

```diff
@@ -282,20 +305,21 @@ struct SourceNode : ExecNode {
 
   void StopProducing(ExecNode* output) override {
     DCHECK_EQ(output, outputs_[0]);
-    {
-      std::unique_lock<std::mutex> lock(mutex_);
-      finished_ = true;
-    }
-    finished_fut_.Wait();
+    StopProducing();
   }
 
-  void StopProducing() override { StopProducing(outputs_[0]); }
+  void StopProducing() override {
+    std::unique_lock<std::mutex> lock(mutex_);
+    stop_requested_ = true;
+  }
+
+  Future<> finished() override { return finished_; }
 
  private:
   std::mutex mutex_;
-  bool finished_{false};
-  int next_batch_index_{0};
-  Future<> finished_fut_ = Future<>::MakeFinished();
+  bool stop_requested_{false};
```

Comment:
Instead of exposing a `StopProducing` you might need to take in a stop token.  If you don't then how will you handle the following...

A pyarrow user runs a to_table on some dataset.  First it has to do inspection, then it has to actually do the scan.  At any point the user might press Ctrl-C to cancel the thing.

We can support it with stop token by setting the Ctrl-C stop token handler and then passing that stop token down to the inspection call and the scan call.

As a minor benefit you can get rid of all the locks here because they are inside the stop token itself.

Reply:
There's an issue (ARROW-12938) for letting a StopToken run a callback since it's needed to support other things like gRPC (which also exposes a cancel method instead taking a cancellable context). But that'd be more work and requires dealing with signal safety in the callback.

Reply:
Replacing with a stop token sounds useful but I'd prefer to handle that in a follow up

Reply:
A follow up sounds fine to me.

Reply:
Filed ARROW-13297.